core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -467,7 +474,7 @@ private[spark] class TaskSchedulerImpl private[scheduler](
       taskState: TaskState,
       reason: TaskFailedReason): Unit = synchronized {
     taskSetManager.handleFailedTask(tid, taskState, reason)
-    if (!taskSetManager.isZombie && taskState != TaskState.KILLED) {
+    if (!taskSetManager.isZombie) {
```

Comment:
I don't think you can get a request storm. The _only_ case this guards against is exactly what you mentioned -- a speculative task that is killed because another attempt succeeded. The number of speculative tasks is always small so this shouldn't be an issue. In comparison, we hit this code path much more often with failed tasks.

If we were to add this check back, then in the task kill API we would have to add a parameter as to whether revive offers should be called. This is substantial added complexity which can be removed if we make this simplification at this site. I'll leave it to @kayousterhout to decide if this is worth it.